### PR TITLE
Add algorithm MAP-Elites-ES to QDax

### PR DESCRIPTION
**Pull request by manon-but-yes**
Related to #81 

This PR adds MAP-Elites-ES to QDax.
The complete MAP-Elites-ES algorithm is implemented as an emitter, including the biased sampling from the repertoire, in `qdax/core/emitters/mees_emitter.py`.

`qdax/utils/sampling.py` also provide necessary functions to perform reevaluations of a given individual. In the case of MAP-Elites-ES this is needed to perform the 30 reevaluations of the final controller generated through ES.

### CODE REVIEW

File: examples/notebooks/mees_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "KeT2NuI_oQvG"
+   },
+   "outputs": [],
+   "source": [
```
Comment (felixchalumeau) [inline examples/notebooks/mees_example.ipynb:247]: Is there no way to use the same scoring function to define the two scoring functions? To avoid the repetition that we have here.

File: qdax/core/emitters/mees_emitter.py [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+    adam_optimizer: bool = True  # if 1 use ADAM, if 0 use SGD
+    learning_rate: float = 0.01
+    learning_rate_decay: float = 1.0  # only applied when no ADAM
+    l2_coefficient: float = 0.02  # coefficient for regularisation
+    novelty_nearest_neighbors: int = 10
+    use_explore: bool = True  # if 0, use only fitness gradient
+
+
+class MEESEmitterState(EmitterState):
+    """Emitter State for the MAP-Elites-ES emitter."""
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:35]: Same here, might be cleaner to put descriptions in the dosctrings :slightly_smiling_face:

File: qdax/core/emitters/mees_emitter.py [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+    random_key: RNGKey
+
+
+class MEESEmitter(Emitter):
+    """
+    Emitter reproducing the MAP-Elites-ES exploit-explore and MAP-Elites-ES exploit
+    algorithms from "Scaling MAP-Elites to Deep Neuroevolution" by Colas et al:
+    https://dl.acm.org/doi/pdf/10.1145/3377930.3390217
+
+    One can choose between the two algorithms by setting the use_explore param:
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:55]: Is there a reason for not having the pure explore one?

File: qdax/core/emitters/mees_emitter.py [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+
+        Args:
+            config
+            total_generations: total number of generations for which the
+                emitter will run, allow to initialise the novelty archive.
+            scoring_fn: used to evaluate the samples for the gradient estimate.
+            num_descriptors: dimension of the descriptors, used to initialise
+                the empty novelty archive.
+
+        Returns: /
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:83]: I think that you can just remove the Returns line when it's void

File: qdax/core/emitters/mees_emitter.py [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+        """
+        self._config = config
+        self._scoring_fn = scoring_fn
+        self._total_generations = total_generations
+        self._num_descriptors = num_descriptors
+
+        # Initialise optimizer
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:90]: I am not a big fan of the way the optimization is either made using an optimizer, either managed internally. Is there no vanilla optimizer that can be imported and used?

File: qdax/core/emitters/mees_emitter.py [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+        """Initializes the emitter state.
+
+        Args:
+            init_genotypes: The initial population.
+            random_key: A random key.
+
+        Returns:
+            The initial state of the MEESEmitter, a new random key.
+        """
+        assert (
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:112]: What does it mean? What is the error? (can you make the error message more explicit?)

File: qdax/core/emitters/mees_emitter.py (hunk 5 above)
Comment (felixchalumeau) [reply qdax/core/emitters/mees_emitter.py:90]: optax.sgd?

File: qdax/core/emitters/mees_emitter.py [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+    use_explore: bool = True  # if 0, use only fitness gradient
+
+
+class MEESEmitterState(EmitterState):
+    """Emitter State for the MAP-Elites-ES emitter."""
+
+    learning_rate: float
+    initial_optimizer_state: optax.OptState  # only used for ADAM
+    optimizer_state: optax.OptState  # only used for ADAM
+    gradient_offspring: Genotype  # Offspring generated through gradient
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:40]: Could it be called offspring?

File: qdax/core/emitters/mees_emitter.py [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+            jax.vmap(partial(distance), in_axes=(None, 0)), in_axes=(0, None)
+        )(descriptors, novelty_archive)
+
+        # Filter distance with empty slot of novelty archive
+        indices = jnp.arange(0, self._total_generations, step=1) < generation_count
+        distances = jax.vmap(lambda distance: jnp.where(indices, distance, jnp.inf))(
+            distances
+        )
+
+        # Find k neirest neighbours
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:213]: Have you experienced any time perf issue with the top_k? (btw, little typo in "nearest")

File: qdax/core/emitters/mees_emitter.py [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+        Returns:
+            The modified emitter state.
+        """
+
+        assert (
+            jax.tree_util.tree_leaves(genotypes)[0].shape[0] == 1
+        ), "ERROR MAP-Elites-ES generates 1 individual per generation (batch_size=1)."
+
+        ############################
+        # Updating novelty archive #
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:366]: I think it would be nicer to have the mechanism hidden in a proper class that implements them. So in this function, we only have to call the method of the class.

Btw, is this way of putting comments a common one?

File: qdax/core/emitters/mees_emitter.py [hunk 10]
```diff
@@ -0,0 +1,613 @@
...
+        # Adding regularisation
+        gradient = jax.tree_map(
+            lambda g, p: g + self._config.l2_coefficient * p,
+            gradient,
+            parent,
+        )
+
+        ######################
+        # Applying gradients #
+
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:583]: It would be nice to avoid having this case distinction here (cf. my comments about the optimizers)

File: qdax/core/emitters/mees_emitter.py [hunk 11]
```diff
@@ -0,0 +1,613 @@
+from dataclasses import dataclass
```
Comment (felixchalumeau) [inline qdax/core/emitters/mees_emitter.py:1]: Generally speaking, I am wandering if it wouldn't be cleaner to having an emitter for explore, another for exploit and have a meta emitter to create the explore/exploit. (not sure it's easy, just wandering)

What do you think about it?

Comment (felixchalumeau) [reply qdax/core/emitters/mees_emitter.py:1]: and obviously, we would take advantage of inheritance to have very small code repetition (or no code repetition at all)

File: examples/notebooks/mees_example.ipynb (hunk 1 above)
Comment (manon-but-yes) [reply examples/notebooks/mees_example.ipynb:234]: Yes it is feasible using the reset_based_scoring_function_brax_envs, I added it.

File: qdax/core/emitters/mees_emitter.py [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+    MapElitesRepertoire,
+    get_cells_indices,
+)
+from qdax.core.emitters.emitter import Emitter, EmitterState
+from qdax.types import Descriptor, ExtraScores, Fitness, Genotype, RNGKey
+
+
+@dataclass
+class MEESConfig:
+    """Configuration for the MAP-Elites-ES emitter."""
```
[the remark this answers is not in the document]
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py]: Done!

File: qdax/core/emitters/mees_emitter.py (hunk 2 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py]: Also done in the last commits!

File: qdax/core/emitters/mees_emitter.py (hunk 3 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py:83]: It was not present in the original paper, do we want to have it?

File: qdax/core/emitters/mees_emitter.py (hunk 4 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py]: Done.

File: qdax/core/emitters/mees_emitter.py (hunk 5 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py:116]: Good point! I changed this.

File: qdax/core/emitters/mees_emitter.py (hunk 6 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py]: I agree the assert is not necessary in the init, I replaced it. However, we still need it for the update: ME-ES has batch-size 1 and would not work with a bigger batch-size. I modified the message in the update() function to try to make it more explicit.

File: qdax/core/emitters/mees_emitter.py (hunk 10 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py:617]: Modified as well with previous comment.

File: qdax/core/emitters/mees_emitter.py (hunk 3 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py:87]: I added it as well, using a use_exploit boolean.

File: qdax/core/emitters/mees_emitter.py (hunk 8 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py]: I did not perform any tests. I was aware of the issue you raised on the speed of this function, but I did not see any other straightforward way to implement this here.

File: qdax/core/emitters/mees_emitter.py (hunk 9 above)
Comment (manon-but-yes) [reply qdax/core/emitters/mees_emitter.py]: I implemented a NoveltyArchive class at the top of the file following your suggestion.
Concerning the comments, I use them as a way to structure the code when it can be divided in a few main steps, I have no idea if it is standardised. I saw you removed them anyway.

**Review by felixchalumeau — APPROVED** [review]
Happy with it.
